nom-sql: Fix LimitOffset parsing/display

nickdavies opened the pull request:
Switch from generic parsing of SQL to having the PostgreSQL and MySQL cases specifically handled and allow for hopefully all of the available syntax choices for `LIMIT` and `OFFSET` for both.

There are 3 tests in `readyset-adapter` which fail with this diff specifically:
- `rewrite::tests::process_query::bare_offset_nonzero`
- `rewrite::tests::process_query::bare_offset_zero`
- `rewrite::tests::process_query::correct_offset_limit`

However they fail because they are trying to use `OFFSET` without `LIMIT` which is not valid for MySQL as far as I can tell. Would you like me to:

-  Change the tests to assert that they fail. I am not sure if that type of test belongs in that crate or not though.
-  Switch the tests to use `Dialect::PostgreSQL` or make that configurable because some use the `LIMIT <offset>, <limit>` syntax from MySQL.
-  Replace the tests with some other test (Recommendations here would be appreciated).
-  Remove the tests that use the invalid syntax (delete `bare_offset_nonzero` and `bare_offset_zero` and remove the specific case that does this from `correct_offset_limit`

I saw from the issue:

> Fully support the postgres limit offset spec and remove the limitations on the Arbitrary derivation that workaround our incomplete parsing so that a round-trip parsing test covers the full behavior.

I was not entirely sure on how to do that step so if there are other tests/cases that I should add please let me know.

Also I was not not entirely sure where certain validation should go and overall structure so feel free to let me know if this isn't idiomatic.

Hopefully resolves #610

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/readysettech/readyset?pullRequest=908) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/readysettech/readyset?pullRequest=908) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/readysettech/readyset?pullRequest=908) it.</sub>

alanamarzoev commented:
Thanks for this @nickdavies! Tagging @lukoktonos and @jasobrown-rs to review.

lukoktonos commented on nom-sql/src/dialect.rs:
```diff
@@ -268,24 +268,136 @@ impl Dialect {
+            let (i, _) = whitespace0(i)?;
+            match self {
+                Dialect::PostgreSQL => {
+                    let (i, limit) = opt(self.limit_only())(i)?;
```
I really like these helpers here--very clean!

nickdavies replied:
Thanks! :)

lukoktonos reviewed (commented):
Thanks for the contribution! This is looking great so far. 

> There are 3 tests in readyset-adapter which fail with this diff specifically...Would you like me to:
> Switch the tests to use Dialect::PostgreSQL or make that configurable because some use the LIMIT <offset>, <limit> syntax from MySQL.

I think this is the right move. The current test seems to be using a mix of Postgres ($1 placeholders) and Mysql (would be ? placeholders) dialects. It would be great to make these tests dialect-aware and have them test the offset/limit differences introduced in this CL. At the very least, we should update the current tests to use a consistent dialect and pass (Postgres, preferably).

> I saw from the issue:
> Fully support the postgres limit offset spec and remove the limitations on the Arbitrary derivation that workaround our incomplete parsing so that a round-trip parsing test covers the full behavior.

> I was not entirely sure on how to do that step so if there are other tests/cases that I should add please let me know.

It looks like you already handled this by updating the `impl Arbitrary...` block to remove some of the limitations we had--there are still the "only generate positive numbers" limitations in place, which makes sense to leave with your TODO comment regarding that. Looks all good there to me.

> Also I was not not entirely sure where certain validation should go and overall structure so feel free to let me know if this isn't idiomatic.

It looks very idiomatic to me! ⭐

nickdavies commented:
Thanks for the review! I updated with your comments. For the tests I have updated them to be more dialect aware but one thing that I was unsure of how you'd like it done is for queries that are either mixed or the test has mixed syntax. Eg: `numbered_where_in_with_auto_params` which has MySQL syntax for the first part and then the other side of the assert is using PostgreSQL syntax:

```
#[test]
fn numbered_where_in_with_auto_params() {
    let (keys, query) = process_and_make_keys_mysql(
        "SELECT * FROM users WHERE x = ? AND y in (?, ?, ?) AND z = 4 AND w = 5 AND q = ?",
        vec![0.into(), 1.into(), 2.into(), 3.into(), 6.into()],
    );   

    assert_eq!(
        query,
        parse_select_statement_postgres(
            "SELECT * FROM users WHERE x = $1 AND y = $2 AND z = $3 AND w = $4 AND q = $5",
        )    
    );   
    ...
```
There are 3 tests like this:
- `rewrite::tests::collapse_where::collapsed_where_dollarsign_placeholders`
- `rewrite::tests::process_query::numbered_auto_parameterized_in`
- `rewrite::tests::process_query::numbered_where_in_with_auto_params`

All those tests pass, however I added `panic!("Should this be allowed to mix mysql and postgres placeholders?");` so I don't miss any spots that have this setup before shipping.

lukoktonos reviewed (approved):
Those tests seem to be sort of doing two separate things--they check that a MySQL or Postgres string converts to a particular rewritten form, and then they also check that a manually crafted, equivalent string of the other dialect converts to the same rewritten form.

I'd say these concerns could probably be separated since we care more about the rewritten forms being correct than we care about the two dialects being equivalent for two particular strings. But since they are passing, I don't see any harm in leaving them as-is for the sake of this PR.

No need to remove the panics on your end--this PR LGTM. I can remove those lines as I `git am` this patch over to our CI system, which will end up merging and syncing with GitHub--you should see this PR marked as "closed" and then the commit will show up on `main` (it will still retain your authorship info).

Thanks for the contribution! 🎉

lukoktonos commented:
Merged in a17fe81f2d20d64dc61282478a7bd59aed2ce86b
